core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -303,26 +326,73 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
   }
 
   /** Install prewarm containers up to the configured requirements for each kind/memory combination. */
-  def backfillPrewarms(init: Boolean) = {
+  def adjustPrewarmedContainer(init: Boolean, scheduled: Boolean): Unit = {
     prewarmConfig.foreach { config =>
+      // Delete unused prewarmed container until minCount is reached
       val kind = config.exec.kind
       val memory = config.memoryLimit
-      val currentCount = prewarmedPool.count {
-        case (_, PreWarmedData(_, `kind`, `memory`, _)) => true //done starting
-        case _                                          => false //started but not finished starting
+
+      val runningCount = prewarmedPool.count {
+        //done starting, and not expired
+        case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if !p.isExpired() => true
+        //started but not finished starting (or expired)
+        case _ => false
       }
       val startingCount = prewarmStartingPool.count(p => p._2._1 == kind && p._2._2 == memory)
-      val containerCount = currentCount + startingCount
-      if (containerCount < config.count) {
+      val currentCount = runningCount + startingCount
+      //determine how many are needed
+      val desiredCount: Int =
+        if (init) config.initialCount
+        else {
+          if (scheduled) {
+            //scheduled/reactive config backfill
+            config.reactive
+              .map(c => getReactiveCold(c, kind, memory).getOrElse(c.minCount)) //reactive -> desired is either cold start driven, or minCount
+              .getOrElse(config.initialCount) //not reactive -> desired is always initial count
+          } else {
+            //normal backfill after removal - make sure at least minCount or initialCount is started
+            config.reactive.map(_.minCount).getOrElse(config.initialCount)
+          }
+        }
+
+      //remove expired
+      if (scheduled) {
+        config.reactive.foreach { reactiveValue =>
+          prewarmedPool
+            .filter { warmInfo =>
+              warmInfo match {
+                case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if p.isExpired() => true
+                case _                                                                  => false
+              }
+            }
+            .drop(reactiveValue.minCount) //keep minCount even if expired
```

Comment:
I think this `.drop` should be removed - the filtering above excludes expired already, so if there is an expired prewarm, it will not be considered in counts, and replaced if desiredCount requires it. So we should always remove all expired prewarms (regardless of minCount). In general this is good, especially if using kubernetes/mesos, so that underused cluster nodes can scale down.

Reply:
Yes, you are right, updated accordingly.